File: apps/csharp/tests/Ch05/PasswordCrackerParallelTests.cs
using GrokkingConcurrency.CSharp.Ch02;
using GrokkingConcurrency.CSharp.Ch05;

namespace GrokkingConcurrency.CSharp.Tests.Ch05;

/// <summary>
/// 並列パスワードクラッカーのテスト
/// </summary>
public class PasswordCrackerParallelTests
{
    [Fact]
    public void ShouldSplitRangeIntoChunks()
    {
        var chunks = PasswordCrackerParallel.GetChunks(4, 4);

        Assert.Equal(4, chunks.Count);
        Assert.Equal(new ChunkRange(0, 2499), chunks[0]);
        Assert.Equal(new ChunkRange(2500, 4999), chunks[1]);
        Assert.Equal(new ChunkRange(5000, 7499), chunks[2]);
        Assert.Equal(new ChunkRange(7500, 9999), chunks[3]);
    }

    [Fact]
    public void ShouldCrackPasswordInChunk()
    {
        var cryptoHash = PasswordCracker.GetCryptoHash("1234");

        var result = PasswordCrackerParallel.CrackChunk(cryptoHash, 4, 0, 9999);

        Assert.Equal("1234", result);
    }

    [Fact]
    public void ShouldReturnNullWhenNotInChunk()
    {
        var cryptoHash = PasswordCracker.GetCryptoHash("5000");

        var result = PasswordCrackerParallel.CrackChunk(cryptoHash, 4, 0, 4999);

        Assert.Null(result);
    }

    [Fact]
    public void ShouldCrackPasswordInParallel()
    {
        var cryptoHash = PasswordCracker.GetCryptoHash("1234");

        var result = PasswordCrackerParallel.CrackPasswordParallel(cryptoHash, 4);

        Assert.Equal("1234", result);
    }

    [Fact]
    public void ShouldCrackLargerPasswordInParallel()
    {
        var cryptoHash = PasswordCracker.GetCryptoHash("54321");

        var result = PasswordCrackerParallel.CrackPasswordParallel(cryptoHash, 5);

        Assert.Equal("54321", result);
    }
}

File: apps/fsharp/tests/PasswordCrackerParallelTests.fs
module PasswordCrackerParallelTests

open Xunit
open GrokkingConcurrency.Ch02.PasswordCracker
open GrokkingConcurrency.Ch05.PasswordCrackerParallel

[<Fact>]
let ``getChunks divides range into equal parts`` () =
    let chunks = getChunks 4 100
    Assert.Equal(4, List.length chunks)
    Assert.Equal(0, (List.head chunks).Start)
    Assert.Equal(100, (List.last chunks).End)

[<Fact>]
let ``getChunks handles uneven division`` () =
    let chunks = getChunks 3 10
    Assert.Equal(3, List.length chunks)
    // Verify all elements are covered
    let covered =
        chunks
        |> List.collect (fun c -> [c.Start .. c.End - 1])
        |> Set.ofList
    Assert.True((Set.ofList [0..9]) = covered)

[<Fact>]
let ``crackPasswordParallel finds password`` () =
    let password = "ab"
    let hash = getCryptoHash password
    Assert.Equal(Some password, crackPasswordParallel hash 2)

[<Fact>]
let ``crackPasswordParallel returns None for non-existent password`` () =
    let hash = "0000000000000000000000000000000000000000000000000000000000000000"
    Assert.Equal(None, crackPasswordParallel hash 2)

[<Fact>]
let ``crackPasswordParallel uses multiple threads`` () =
    let password = "zz"
    let hash = getCryptoHash password
    let result = crackPasswordParallel hash 2
    Assert.Equal(Some password, result)
